org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -373,5 +405,29 @@ protected void doClean(ToBeBuilt toBeBuilt, IProgressMonitor monitor) throws Cor
 			progress.worked(1);
 		}
 	}
-	
+
+	@Override
+	public ISchedulingRule getRule(int kind, Map<String, String> args) {
+		switch (getSchedulingOption()) {
+			case NULL: return null;
+			case WORKSPACE: return getProject().getWorkspace().getRoot();
+			case PROJECT: return getProject();
+			case ALL_XTEXT_PROJECTS: return new MultiRule(Arrays.stream(
+				getProject().getWorkspace().getRoot().getProjects())
+				.filter(XtextProjectHelper::hasNature)
+				.toArray(ISchedulingRule[]::new));
+			default: throw new IllegalArgumentException();
+		}
+	}
+
+	/**
+	 * @since 2.16
+	 */
+	private SchedulingOption getSchedulingOption () {
+		String schedulingRuleName = preferenceStore.getString(PREF_SCHEDULING_RULE);
+		if (schedulingRuleName.isEmpty()) {
+			return SchedulingOption.WORKSPACE;
+		}
+		return SchedulingOption.valueOf(preferenceStore.getString(PREF_SCHEDULING_RULE));
```

Comment:
nitpicky: why not use schedulingRuleName

Reply:
Is it really just a name? It is more a strategy.

Reply:
got me wrong. why calling  SchedulingOption.valueOf(preferenceStore.getString(PREF_SCHEDULING_RULE)); twice

Reply:
OK, makes sense. I thought you were referring to the preference's name. Will update.

Reply:
updated the pr